Add TrackInfo/AudioInfo/VideoInfo serialisers

jyavenard opened the pull request:
#### b5ee8dd92ca787894684fb908fe5ccf9318bbef3
<pre>
Add TrackInfo/AudioInfo/VideoInfo serialisers
<a href="https://bugs.webkit.org/show_bug.cgi?id=299169">https://bugs.webkit.org/show_bug.cgi?id=299169</a>
<a href="https://rdar.apple.com/160929565">rdar://160929565</a>

Reviewed by Youenn Fablet.

We can also remove the unused RemoteAudioInfo/RemoteVideoInfo object
they are unusused and would have redundant following this commit.

No change in observable behaviours. The serialisers will be used in a
follow-up change.

* Source/WebCore/platform/TrackInfo.cpp:
(WebCore::TrackInfo::createWithVariant):
* Source/WebCore/platform/TrackInfo.h:
(WebCore::TrackInfo::TrackInfo):
(WebCore::VideoInfo::create):
(WebCore::VideoInfo::VideoInfo):
(WebCore::AudioInfo::AudioInfo):
(WebCore::AudioInfo::create):
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/media/RemoteTrackInfo.h: Removed.
* Source/WebKit/GPUProcess/media/RemoteTrackInfo.serialization.in: Removed.
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/300399@main">https://commits.webkit.org/300399@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/51ddb3c0318754034158313c28baa29d455f890a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122361 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42066 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32753 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128947 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74460 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/caf958d6-32e6-4c8b-b5a2-f8dd4e46245c) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124237 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42786 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50660 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93004 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/61802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6c79a4df-ce43-4fec-bd89-4f59a1b1e327) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125313 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34121 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109560 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73663 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/96e5cbaf-70c4-48be-9a1f-7d8c36abdbf0) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/121718 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33110 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27719 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72438 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103704 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27927 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131689 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49302 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37517 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101554 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49676 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105780 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101425 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25736 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46800 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24929 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/46063 "Build was cancelled. Recent messages:Printed configuration") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49160 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54902 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48627 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51977 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50309 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/TrackInfo.cpp:
```diff
@@ -44,4 +44,11 @@ String convertEnumerationToString(TrackInfoTrackType type)
     return values[static_cast<size_t>(type)];
 }
 
+Ref<TrackInfo> TrackInfo::createWithVariant(Variant<Ref<AudioInfo>, Ref<VideoInfo>> variant)
```
```suggestion
Ref<TrackInfo> TrackInfo::fromVariant(Variant<Ref<AudioInfo>, Ref<VideoInfo>>&& variant)
```

youennf commented on Source/WebCore/platform/TrackInfo.h:
```diff
@@ -83,8 +87,19 @@ struct TrackInfo : public ThreadSafeRefCounted<TrackInfo> {
+    friend struct IPC::ArgumentCoder<AudioInfo, void>;
+    friend struct IPC::ArgumentCoder<VideoInfo, void>;
+    static Ref<TrackInfo> createWithVariant(Variant<Ref<AudioInfo>, Ref<VideoInfo>>);
     const TrackType m_type { TrackType::Unknown };
```
I would make TrackInfo and relevant classes instead of structs.

jyavenard replied:
That's out of the scope of this patch, and lots of code rely on those to be structs, directly assigning the members

youennf commented on Source/WebCore/platform/TrackInfo.h:
```diff
@@ -103,6 +118,24 @@ struct VideoInfo : public TrackInfo {
+
+    // Used by IPC generator
+    friend struct IPC::ArgumentCoder<VideoInfo, void>;
+    static Ref<VideoInfo> create(FourCC codecName, const String& codecString, WebCore::TrackID trackID, FloatSize size, FloatSize displaySize, uint8_t bitDepth, PlatformVideoColorSpace colorSpace, RefPtr<SharedBuffer> atomData)
```
```suggestion
    static Ref<VideoInfo> create(FourCC codecName, const String& codecString, WebCore::TrackID trackID, FloatSize size, FloatSize displaySize, uint8_t bitDepth, PlatformVideoColorSpace colorSpace, RefPtr<SharedBuffer>&& atomData)
```

youennf commented on Source/WebCore/platform/TrackInfo.h:
```diff
@@ -103,6 +118,24 @@ struct VideoInfo : public TrackInfo {
+        return adoptRef(*new VideoInfo(codecName, codecString, trackID, size, displaySize, bitDepth, colorSpace, WTFMove(atomData)));
+    }
+
+    VideoInfo(FourCC codecName, const String& codecString, WebCore::TrackID trackID, FloatSize size, FloatSize displaySize, uint8_t bitDepth, PlatformVideoColorSpace colorSpace, RefPtr<SharedBuffer> atomData)
```
```suggestion
    VideoInfo(FourCC codecName, const String& codecString, WebCore::TrackID trackID, FloatSize size, FloatSize displaySize, uint8_t bitDepth, PlatformVideoColorSpace colorSpace, RefPtr<SharedBuffer>&& atomData)
```

youennf commented on Source/WebCore/platform/TrackInfo.h:
```diff
@@ -122,7 +155,25 @@ struct AudioInfo : public TrackInfo {
+
+    // Used by IPC generator
+    friend struct IPC::ArgumentCoder<AudioInfo, void>;
+    static Ref<AudioInfo> create(FourCC codecName, const String& codecString, TrackID trackID, uint32_t rate, uint32_t channels, uint32_t framesPerPacket, uint8_t bitDepth, RefPtr<SharedBuffer> cookieData)
```
```suggestion
    static Ref<AudioInfo> create(FourCC codecName, const String& codecString, TrackID trackID, uint32_t rate, uint32_t channels, uint32_t framesPerPacket, uint8_t bitDepth, RefPtr<SharedBuffer>&& cookieData)
```

youennf commented on Source/WebCore/platform/TrackInfo.h:
```diff
@@ -122,7 +155,25 @@ struct AudioInfo : public TrackInfo {
+        return adoptRef(*new AudioInfo(codecName, codecString, trackID, rate, channels, framesPerPacket, bitDepth, WTFMove(cookieData)));
+    }
+
+    AudioInfo(FourCC codecName, const String& codecString, TrackID trackID, uint32_t rate, uint32_t channels, uint32_t framesPerPacket, uint8_t bitDepth, RefPtr<SharedBuffer> cookieData)
```
```suggestion
    AudioInfo(FourCC codecName, const String& codecString, TrackID trackID, uint32_t rate, uint32_t channels, uint32_t framesPerPacket, uint8_t bitDepth, RefPtr<SharedBuffer>&& cookieData)
```

jyavenard committed:
Add TrackInfo/AudioInfo/VideoInfo serialisers
https://bugs.webkit.org/show_bug.cgi?id=299169
rdar://160929565

Reviewed by Youenn Fablet.

We can also remove the unused RemoteAudioInfo/RemoteVideoInfo object
they are unusused and would have redundant following this commit.

No change in observable behaviours. The serialisers will be used in a
follow-up change.

* Source/WebCore/platform/TrackInfo.cpp:
(WebCore::TrackInfo::createWithVariant):
* Source/WebCore/platform/TrackInfo.h:
(WebCore::TrackInfo::TrackInfo):
(WebCore::VideoInfo::create):
(WebCore::VideoInfo::VideoInfo):
(WebCore::AudioInfo::AudioInfo):
(WebCore::AudioInfo::create):
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/media/RemoteTrackInfo.h: Removed.
* Source/WebKit/GPUProcess/media/RemoteTrackInfo.serialization.in: Removed.
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/300399@main
